apollo-ios/Sources/Apollo/FieldPolicyEvaluator.swift

```diff
@@ -0,0 +1,190 @@
+      }
+    }
+
+    var fixedParts = [String?](repeating: nil, count: parsed.count)
+    var listIndex: Int? = nil
+    var listValues: [String] = []
+
+    for (i, pk) in parsed.enumerated() {
+      guard let argVal = arguments[pk.name] else {
+        return nil
+      }
+
+      guard let resolved = argVal.resolveValue(keyPath: pk.path, variables: variables), !resolved.isEmpty else {
+        return nil
+      }
+      
+      if resolved.count > 1 {
+        listIndex = i
+        listValues = resolved
+      } else {
+        guard let value = resolved.first else {
+          return nil
+        }
+        fixedParts[i] = value
+      }
+    }
+
+    if let idx = listIndex {
+      var keys: [CacheKeyInfo] = []
+      keys.reserveCapacity(listValues.count)
+      for item in listValues {
+        var parts = [String]()
+        parts.reserveCapacity(parsed.count)
+        for j in 0..<parsed.count {
+          if j == idx {
+            parts.append(item)
+          } else if let v = fixedParts[j] {
+            parts.append(v)
+          } else {
+            return nil
+          }
+        }
+        keys.append(CacheKeyInfo(id: parts.joined(separator: "+")))
+      }
+      return .list(keys)
+    } else {
+      let parts = fixedParts.compactMap { $0 }
+      guard parts.count == parsed.count else {
+        return nil
+      }
+      return .single(CacheKeyInfo(id: parts.joined(separator: "+")))
+    }
+  }
+  
+}
+
+extension ScalarType {
+  fileprivate var cacheKeyComponentStringValue: String {
+    switch self {
+    case let strVal as String:
+      return strVal
+    case let boolVal as Bool:
+      return boolVal ? "true" : "false"
+    case let intVal as Int:
+      return String(intVal)
+    case let doubleVal as Double:
+      return String(doubleVal)
+    case let floatVal as Float:
+      return String(floatVal)
+    default:
+      return String(describing: self)
+    }
+  }
+}
+
+extension JSONValue {
+  fileprivate func cacheKeyComponentStringValue(keyPath: [String]? = nil) -> [String]? {
+    switch self {
+    case let strVal as String:
+      return [strVal]
+    case let boolVal as Bool:
+      return boolVal ? ["true"] : ["false"]
+    case let intVal as Int:
+      return [String(intVal)]
+    case let doubleVal as Double:
+      return [String(doubleVal)]
+    case let floatVal as Float:
+      return [String(floatVal)]
+    case let arrVal as [JSONValue]:
+      let values: [String] = arrVal.compactMap { $0.cacheKeyComponentStringValue()?.first }
+      guard !values.isEmpty else { return nil }
+      return values
+    case let objVal as JSONObject:
+      guard let keyPath, !keyPath.isEmpty else { return nil }
+      guard let targetValue = objVal.walk(path: keyPath[...]) else { return nil }
+      return targetValue.cacheKeyComponentStringValue()
+    default:
+      return [String(describing: self)]
+    }
+  }
+}
+
+extension JSONObject {
+  fileprivate func walk(
```

Comment:
I don't know what this `walk` function does. The name needs to be more descriptive of the intent here.